rest/config_test.go

```diff
@@ -2925,3 +2926,146 @@ func makeScopesConfigWithDefault(scopeName string, collections []string) *Scopes
 	scopesConfig := makeScopesConfig(scopeName, collections)
 	return &scopesConfig
 }
+
+// TestInvalidDbConfigNoLongerPresentInBucket:
+//   - Create rest tester with large config poll interval
+//   - Create valid db
+//   - Alter config in bucket to make it invalid
+//   - Force config poll, assert it is picked up as invalid db config
+//   - Delete the invalid db config form the bucket
+//   - Force config poll reload and assert the invalid db is cleared
+func TestInvalidDbConfigNoLongerPresentInBucket(t *testing.T) {
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("test only works with CBS, requires bootstrap connection")
+	}
+	rt := NewRestTester(t, &RestTesterConfig{
+		CustomTestBucket: base.GetTestBucket(t),
+		PersistentConfig: true,
+		MutateStartupConfig: func(config *StartupConfig) {
+			// configure the interval time to not run
+			config.Bootstrap.ConfigUpdateFrequency = base.NewConfigDuration(10 * time.Minute)
+		},
+		DatabaseConfig: nil,
+	})
+	defer rt.Close()
+	realBucketName := rt.CustomTestBucket.GetName()
+	ctx := base.TestCtx(t)
+	const dbName = "db1"
+
+	// create db with correct config
+	dbConfig := rt.NewDbConfig()
+	resp := rt.CreateDatabase(dbName, dbConfig)
+	RequireStatus(t, resp, http.StatusCreated)
+
+	// wait for db to come online
+	require.NoError(t, rt.WaitForDBOnline())
+
+	// grab the persisted db config from the bucket
+	databaseConfig := DatabaseConfig{}
+	_, err := rt.ServerContext().BootstrapContext.GetConfig(rt.Context(), realBucketName, rt.ServerContext().Config.Bootstrap.ConfigGroupID, "db1", &databaseConfig)
+	require.NoError(t, err)
+
+	// update the persisted config to a fake bucket name
+	newBucketName := "fakeBucket"
+	_, err = rt.UpdatePersistedBucketName(&databaseConfig, &newBucketName)
+	require.NoError(t, err)
+
+	// force reload of configs from bucket
+	rt.ServerContext().ForceDbConfigsReload(t, ctx)
+
+	// assert the config is picked as invalid db config
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		invalidDatabases := rt.ServerContext().AllInvalidDatabaseNames(t)
+		assert.Equal(c, 1, len(invalidDatabases))
+		assert.Equal(c, 0, len(rt.ServerContext().dbConfigs))
+	}, time.Second*10, time.Millisecond*100)
+
+	// remove the invalid config from the bucket
+	rt.RemoveDbConfigFromBucket(dbName, realBucketName)
+
+	// force reload of configs from bucket
+	rt.ServerContext().ForceDbConfigsReload(t, ctx)
+
+	// assert the config is removed from tracking
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		invalidDatabases := rt.ServerContext().AllInvalidDatabaseNames(t)
+		assert.Equal(c, 0, len(invalidDatabases))
```

Comment:
From linked integration test. Does this need more code backporting?

```go
2024-08-13T02:27:23.094Z [WRN] t:TestInvalidDbConfigNoLongerPresentInBucket b:sg_int_2_1723515670481772940 Must repair invalid database config for "db1" for it to be usable! Mismatched buckets (config bucket: "fakeBucket", actual bucket: "sg_int_2_1723515670481772940") -- rest.(*invalidDatabaseConfigs).addInvalidDatabase() at config.go:348
2024-08-13T02:27:23.094Z [INF] t:TestInvalidDbConfigNoLongerPresentInBucket b:sg_int_2_1723515670481772940 Closing db /db1 (bucket "sg_int_2_1723515670481772940")
2024/08/13 02:27:23 TEST: Memory high water mark increased to 2199.86 MB (heap: 2195.20 MB stack: 4.66 MB)
2024/08/13 02:27:28 TEST: Memory high water mark increased to 2233.84 MB (heap: 2229.38 MB stack: 4.47 MB)
2024/08/13 02:27:33 TEST: Memory high water mark increased to 2241.92 MB (heap: 2237.45 MB stack: 4.47 MB)
    config_test.go:2990: 
        	Error Trace:	/home/ec2-user/workspace/SyncGateway-Integration/rest/config_test.go:2992
        	            				/home/ec2-user/tools/org.jenkinsci.plugins.golang.GolangInstallation/1.22.5/src/runtime/asm_amd64.s:1695
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 1
    config_test.go:2990: 
        	Error Trace:	/home/ec2-user/workspace/SyncGateway-Integration/rest/config_test.go:2990
        	Error:      	Condition never satisfied
        	Test:       	TestInvalidDbConfigNoLongerPresentInBucket2024-08-13T02:27:33.291Z [INF] t:TestNotFoundOnInvalidDatabase Logging stats with frequency: &{1m0s}
```

Reply:
This test looks to be from CBG-4087